jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -67,68 +69,154 @@ public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity)
     {
+        this(minCapacity, factor, maxCapacity, maxQueueLength, -1, -1);
+    }
+
+    /**
+     * Creates a new ArrayByteBufferPool with the given configuration.
+     *
+     * @param minCapacity the minimum ByteBuffer capacity
+     * @param factor the capacity factor
+     * @param maxCapacity the maximum ByteBuffer capacity
+     * @param maxQueueLength the maximum ByteBuffer queue length
+     * @param maxHeapMemory the max heap memory in bytes
+     * @param maxDirectMemory the max direct memory in bytes
+     */
+    public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity, int maxQueueLength, long maxHeapMemory, long maxDirectMemory)
+    {
+        super(factor, maxQueueLength, maxHeapMemory, maxDirectMemory);
+
+        factor = getCapacityFactor();
         if (minCapacity <= 0)
             minCapacity = 0;
-        if (factor <= 0)
-            factor = 1024;
         if (maxCapacity <= 0)
             maxCapacity = 64 * 1024;
         if ((maxCapacity % factor) != 0 || factor >= maxCapacity)
             throw new IllegalArgumentException("The capacity factor must be a divisor of maxCapacity");
         _minCapacity = minCapacity;
-        _factor = factor;
 
         int length = maxCapacity / factor;
         _direct = new ByteBufferPool.Bucket[length];
         _indirect = new ByteBufferPool.Bucket[length];
-
-        int capacity = 0;
-        for (int i = 0; i < _direct.length; ++i)
-        {
-            capacity += _factor;
-            _direct[i] = new ByteBufferPool.Bucket(this, capacity, maxQueueLength);
-            _indirect[i] = new ByteBufferPool.Bucket(this, capacity, maxQueueLength);
-        }
     }
 
     @Override
     public ByteBuffer acquire(int size, boolean direct)
     {
-        ByteBufferPool.Bucket bucket = bucketFor(size, direct);
+        int capacity = size < _minCapacity ? size : (bucketFor(size) + 1) * getCapacityFactor();
+        ByteBufferPool.Bucket bucket = bucketFor(size, direct, null);
         if (bucket == null)
-            return newByteBuffer(size, direct);
-        return bucket.acquire(direct);
+            return newByteBuffer(capacity, direct);
+        ByteBuffer buffer = bucket.acquire();
+        if (buffer == null)
+            return newByteBuffer(capacity, direct);
+        decrementMemory(buffer);
+        return buffer;
     }
 
     @Override
     public void release(ByteBuffer buffer)
     {
-        if (buffer != null)
+        if (buffer == null)
+            return;
+        boolean direct = buffer.isDirect();
+        ByteBufferPool.Bucket bucket = bucketFor(buffer.capacity(), direct, this::newBucket);
+        if (bucket != null)
         {
-            ByteBufferPool.Bucket bucket = bucketFor(buffer.capacity(), buffer.isDirect());
-            if (bucket != null)
-                bucket.release(buffer);
+            bucket.release(buffer);
+            incrementMemory(buffer);
```

Comment:
maybe increment memory and release excess before doing the actual release?  Again not exactly sure why... just a feeling.

Reply:
By releasing the buffer early, I give it a chance to be acquired _before_ I do the memory work.